### PR TITLE
Fix an issue with ecma collections in the Promise implementation

### LINKED ISSUE #2486: Assertion 'ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDDATA || ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR' failed in ecma_is_property_lcached
###### Jerry version:

```
Checked revision: 300e40ba
Build command: tools/build.py --debug --logging=on --profile=es2015-subset
```

###### OS:

```
Ubuntu 18.04, x86_64
```

###### Test case:

```javascript
Object.setPrototypeOf(Math, Int32Array);
for (var i = 0; i < 200; i++) { 
    Promise.race([, [,] % {}]).then(); 
}

```

The above test case can cause 3 different assertion failures if running multiple times:

###### Failure 1:

```
ICE: Assertion 'ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDDATA || ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR' failed at jerryscript/jerry-core/ecma/base/ecma-helpers.c(ecma_is_property_lcached):1258.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Failure 2:

```
ICE: Assertion 'ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p)' failed at jerryscript/jerry-core/ecma/base/ecma-helpers.c(ecma_find_named_property):627.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Failure 3:

```
ICE: Assertion 'ecma_is_value_boolean (handler) || ecma_op_is_callable (handler)' failed at jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c(ecma_process_promise_reaction_job):158.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### LINKED ISSUE #2506: Assertion 'ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p)' failed in ecma_find_named_property
###### Jerry version:

```
Checked revision: d270f82e
Build command: ./tools/build.py --clean --debug --error-messages=on --profile=es2015-subset --logging=on
```

###### OS:

```
Ubuntu 18.04, x86_64
```

###### Test case:

```javascript
print(Promise.race([,,,,,,,,,,,,,,,this,,this,,,,,,{ },function(){}])
	.then(function(){$('The promise should be rejected')},function(){$.id_1(escape,URIError)})
	.then(-Math.PI/0)<+RangeError);Date.UTC("",0);
new Date( ).getHours(-{ });
$.$("arg1,arg2,arg3","arg1,arg2",$,$);
$.setYear
```

###### Backtrace:

```
false
Script Error: ReferenceError: $ is not defined
ICE: Assertion 'ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p)' failed at jerryscript/jerry-core/ecma/base/ecma-helpers.c(ecma_find_named_property):627.
Error: ERR_FAILED_INTERNAL_ASSERTION

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff7a24801 in __GI_abort () at abort.c:79
#2  0x000055555555efae in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION)
    at jerryscript/jerry-port/default/default-fatal.c:71
#3  0x0000555555589c30 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION)
    at jerryscript/jerry-core/jrt/jrt-fatals.c:63
#4  0x0000555555589c82 in jerry_assert_fail (assertion=0x5555555e02e8 "ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p)", 
    file=0x5555555e1e18 "jerryscript/jerry-core/ecma/base/ecma-helpers.c", 
    function=0x5555555ceab0 <__func__.4374.lto_priv.126> "ecma_find_named_property", line=627)
    at jerryscript/jerry-core/jrt/jrt-fatals.c:87
#5  0x00005555555b0117 in ecma_find_named_property (obj_p=0x5555558f5178 <jerry_global_heap+3528>, name_p=0x75)
    at jerryscript/jerry-core/ecma/base/ecma-helpers.c:627
#6  0x00005555555a4abb in ecma_op_object_find_own (base_value=3531, object_p=0x5555558f5178 <jerry_global_heap+3528>, 
    property_name_p=0x75) at jerryscript/jerry-core/ecma/operations/ecma-objects.c:508
#7  0x00005555555a4e77 in ecma_op_object_get (object_p=0x5555558f5178 <jerry_global_heap+3528>, property_name_p=0x75)
    at jerryscript/jerry-core/ecma/operations/ecma-objects.c:683
#8  0x00005555555a19f1 in ecma_process_promise_reaction_job (obj_p=0x5555558f4478 <jerry_global_heap+200>)
    at jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c:154
#9  0x00005555555a1eb9 in ecma_process_all_enqueued_jobs ()
    at jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c:343
#10 0x00005555555bf989 in jerry_run_all_enqueued_jobs () at jerryscript/jerry-core/api/jerry.c:570
#11 0x00005555555be6a0 in main (argc=3, argv=0x7fffffffdbf8) at jerryscript/jerry-main/main-unix.c:863
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
Fixes #2486
Fixes #2506

Co-authored-by: Robert Fancsik frobert@inf.u-szeged.hu
JerryScript-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-gc.c [hunk 1]
```diff
@@ -231,16 +231,30 @@ ecma_gc_mark_promise_object (ecma_extended_object_t *ext_object_p) /**< extended
...
-    ecma_gc_set_object_visited (ecma_get_object_from_value (*ecma_value_p));
-    ecma_value_p = ecma_collection_iterator_next (ecma_value_p);
+    if (ecma_is_value_object (*ecma_value_p))
+    {
+      ecma_gc_set_object_visited (ecma_get_object_from_value (*ecma_value_p));
+      ecma_value_p = ecma_collection_iterator_next (ecma_value_p);
+    }
+    else
+    {
+      break;
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c]: Aren't these callables? If yes, why don't we simply prevent adding non=objects to this collection?

File: jerry-core/ecma/base/ecma-gc.c [hunk 2]
```diff
@@ -231,16 +231,30 @@ ecma_gc_mark_promise_object (ecma_extended_object_t *ext_object_p) /**< extended
...
```
Comment (akosthekiss) [inline jerry-core/ecma/base/ecma-gc.c]: My two cents here: isn't this an overly complicated way of writing
```c
while (ecma_value_p != NULL && ecma_is_value_object (*ecma_value_p))
```
? The `if` in the `while` is only justified if there is no `else` with a `break`, IMO. (I think that's what @zherczeg was asking.)

File: jerry-core/ecma/operations/ecma-promise-object.c [hunk 3]
```diff
@@ -176,16 +176,12 @@ ecma_reject_promise (ecma_value_t promise, /**< promise */
   ecma_promise_set_result (obj_p, ecma_copy_value_if_not_object (reason));
   ecma_promise_object_t *promise_p = (ecma_promise_object_t *) obj_p;
 
-  /* GC can be triggered by ecma_new_values_collection so freeing the collection
-     first and creating a new one might cause a heap after use event. */
```
Comment (akosthekiss) [inline jerry-core/ecma/operations/ecma-promise-object.c]: Is this change needed / valid? The ASAN job on Travis CI is complaining exactly about a heap-use-after-free.

Comment (DanielBallaSZTE) [reply jerry-core/ecma/operations/ecma-promise-object.c]: Sadly I can't reproduce the same issue with the same build..

**Comment by DanielBallaSZTE** [comment]
I cannot reproduce any of the travis failures with either version of the patch. 
 

- The first version which removes the `heap after use` fix from the `ecma-promise-object.c`, fails on the ASAN job. However, only on Travis. Having tested it on several machines with different OS / gcc versions, no failures could be reproduced at all.
- The second version which leaves it as it is on the master, causes two jobs to fail on Travis, however none of them could I reproduce locally.

**Comment by akosthekiss** [comment]
Analysis is coming (this gona be long).

So, first on the variant when `ecma_reject_promise` **is changed**. Here is what Travis CI is reporting, if the "GC can be triggered by ecma_new_values_collection so freeing the collection first and creating a new one might cause a heap after use event" part in `ecma_reject_promise` is tampered with:

```
[  58/ 395]FAIL (1): tests/jerry/es2015/regression-test-issue-1881.js

=================================================================
==16738==ERROR: AddressSanitizer: heap-use-after-free on address 0xf4d00698 at pc 0x08050b1d bp 0xffce4ba8 sp 0xffce4b9c
READ of size 4 at 0xf4d00698 thread T0
    #0 0x8050b1c in ecma_collection_iterator_init /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/base/ecma-helpers-values-collection.c:163
    #1 0x8050fd6 in ecma_gc_mark_promise_object /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/base/ecma-gc.c:230
    #2 0x8050fd6 in ecma_gc_mark /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/base/ecma-gc.c:339
    #3 0x805a823 in ecma_gc_run /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/base/ecma-gc.c:863
    #4 0x807b6cd in jmem_run_free_unused_memory_callbacks /home/travis/build/jerryscript-project/jerryscript/jerry-core/jmem/jmem-allocator.c:267
    #5 0x807b766 in jmem_heap_gc_and_alloc_block.lto_priv.56 /home/travis/build/jerryscript-project/jerryscript/jerry-core/jmem/jmem-heap.c:317
    #6 0x805177b in jmem_heap_alloc_block /home/travis/build/jerryscript-project/jerryscript/jerry-core/jmem/jmem-heap.c:366
    #7 0x805177b in jmem_pools_alloc /home/travis/build/jerryscript-project/jerryscript/jerry-core/jmem/jmem-poolman.c:104
    #8 0x805177b in ecma_new_values_collection /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/base/ecma-helpers-values-collection.c:43
    #9 0x808263c in ecma_reject_promise /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-promise-object.c:183
    #10 0x8082bf0 in ecma_promise_reject_handler /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-promise-object.c:251
    #11 0x80727b2 in ecma_op_function_call /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-function-object.c:589
    #12 0x807361e in ecma_process_promise_reaction_job /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c:203
    #13 0x805c7d5 in ecma_process_all_enqueued_jobs /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c:343
    #14 0x805c7d5 in jerry_run_all_enqueued_jobs /home/travis/build/jerryscript-project/jerryscript/jerry-core/api/jerry.c:571
    #15 0x804a5d8 in main /home/travis/build/jerryscript-project/jerryscript/jerry-main/main-unix.c:863
    #16 0xf7007ad2 in __libc_start_main (/lib32/libc.so.6+0x19ad2)
    #17 0x804ad19  (/home/travis/build/jerryscript-project/jerryscript/build/tests/jerry_tests-es2015_subset-debug/local/bin/jerry+0x804ad19)

0xf4d00698 is located 8 bytes inside of 16-byte region [0xf4d00690,0xf4d006a0)
freed by thread T0 here:
    #0 0xf724c14e in free (/usr/lib32/libasan.so.2+0x9214e)
    #1 0x8052f7b in jmem_heap_free_block /home/travis/build/jerryscript-project/jerryscript/jerry-core/jmem/jmem-heap.c:490
    #2 0x8052f7b in ecma_free_values_collection /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/base/ecma-helpers-values-collection.c:62
    #3 0x808260f in ecma_reject_promise /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-promise-object.c:180
    #4 0x8082bf0 in ecma_promise_reject_handler /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-promise-object.c:251
    #5 0x80727b2 in ecma_op_function_call /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-function-object.c:589
    #6 0x807361e in ecma_process_promise_reaction_job /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c:203
    #7 0x805c7d5 in ecma_process_all_enqueued_jobs /home/travis/build/jerryscript-project/jerryscript/jerry-core/ecma/operations/ecma-jobqueue.c:343
    #8 0x805c7d5 in jerry_run_all_enqueued_jobs /home/travis/build/jerryscript-project/jerryscript/jerry-core/api/jerry.c:571
    #9 0x804a5d8 in main /home/travis/build/jerryscript-project/jerryscript/jerry-main/main-unix.c:863
    #10 0xf7007ad2 in __libc_start_main (/lib32/libc.so.6+0x19ad2)
```

These logs show that `ecma_reject_promise` calls `ecma_free_values_collection` (freed#3 to freed#2), and then calls `ecma_new_values_collection` (read#9 to read#8), which triggers the GC (read#5 and up), which tries to access the previously freed collection (in read#0). So, it seems to me that the warning comment on why not to rewrite this part of the code is quite accurate and right. Thus, I'd still suggest not to touch this code.

On the failure when `ecma_reject_promise` **is not changed**. The logs are then (in a somewhat shortened form):
```
/home/travis/build/jerryscript-project/jerryscript/tools/build.py \
--profile=es2015-subset \
--debug \

[  70/ 396]FAIL (120): tests/jerry/es2015/regression-test-issue-2486.js
```
And this is something that I could reproduce out of the box by
```sh
$ git fetch origin pull/2518/head:pr2518
$ git checkout pr2518
$ tools/build.py --profile=es2015-subset --debug
$ build/bin/jerry tests/jerry/es2015/regression-test-issue-2486.js
$ echo $?
120
```
With a slightly enhanced command line, the ICE is also printed:
```sh
$ tools/build.py --profile=es2015-subset --debug --logging=on
$ build/bin/jerry tests/jerry/es2015/regression-test-issue-2486.js
ICE: Assertion 'ecma_is_value_boolean (handler) || ecma_op_is_callable (handler)' failed at jerry-core/ecma/operations/ecma-jobqueue.c(ecma_process_promise_reaction_job):158.
Error: ERR_FAILED_INTERNAL_ASSERTION
$ echo $?
120
```
This is failure 3 from the issue report. This tells me that this PR is not fixing issue 2486.

And one more thing, on the rest of the PR. If I use the above build, the other new test (from issue 2506) is just failing fine (i.e., without an assertion error):
```sh
$ build/bin/jerry tests/jerry/fail/regression-test-issue-2506.js 
false
Script Error: ReferenceError
$ echo $?
1
```
However, if I build jerry as reported in the issue (i.e., with error messages enabled), then the ICE is still there:
```sh
$ tools/build.py --clean --debug --error-messages=on --profile=es2015-subset --logging=on
$ build/bin/jerry tests/jerry/fail/regression-test-issue-2506.js 
false
Script Error: ReferenceError: $ is not defined
ICE: Assertion 'ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p)' failed at jerry-core/ecma/base/ecma-helpers.c(ecma_find_named_property):629.
Error: ERR_FAILED_INTERNAL_ASSERTION
$ echo $?
120
```
This is the error from the issue report. This tells me that this PR is not fixing 2506 either, unfortunately, and the proposed code changes are unnecessary in their current form.

(Unfortunately, this may also hint at the test framework not being able to catch issue 2506 with its current build settings.)

**Comment by DanielBallaSZTE** [comment]
@akosthekiss Thank you for your explanation, having read your comment definitely highlighted for me that this issue needs some more in-depth investigation to be fixed.
As for now, I might close this PR, until I find a correct solution, and reopen it later. (Only of course, if you are fine with this.)

**Pull request closed without merging**